consensus/polybft/extra.go

```diff
@@ -0,0 +1,390 @@
+func (i *Extra) MarshalRLPWith(ar *fastrlp.Arena) *fastrlp.Value {
+	vv := ar.NewArray()
+
+	// Validators
+	if i.Validators == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Validators.MarshalRLPWith(ar))
+	}
+
+	// Seal
+	if len(i.Seal) == 0 {
+		vv.Set(ar.NewNull())
+	} else {
+		vv.Set(ar.NewBytes(i.Seal))
+	}
+
+	// ParentSeal
+	if i.Parent == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Parent.MarshalRLPWith(ar))
+	}
+
+	// CommittedSeal
+	if i.Committed == nil {
+		vv.Set(ar.NewNullArray())
+	} else {
+		vv.Set(i.Committed.MarshalRLPWith(ar))
+	}
+
+	return vv
+}
+
+// UnmarshalRLP defines the unmarshal function wrapper for Extra
+func (i *Extra) UnmarshalRLP(input []byte) error {
+	return fastrlp.UnmarshalRLP(input, i)
+}
+
+// UnmarshalRLPWith defines the unmarshal implementation for Extra
+func (i *Extra) UnmarshalRLPWith(v *fastrlp.Value) error {
+	elems, err := v.GetElems()
+	if err != nil {
+		return err
+	}
+
+	if num := len(elems); num != 4 {
+		return fmt.Errorf("not enough elements to decode extra, expected 4 but found %d", num)
+	}
+
+	// Validators
+	{
+		if elems[0].Elems() > 0 {
+			i.Validators = &ValidatorSetDelta{}
+			if err := i.Validators.UnmarshalRLPWith(elems[0]); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Seal
+	{
+		if elems[1].Len() > 0 {
+			if i.Seal, err = elems[1].GetBytes(i.Seal); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Parent
+	{
+		if elems[2].Elems() > 0 {
+			i.Parent = &Signature{}
+			if err := i.Parent.UnmarshalRLPWith(elems[2]); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Committed
+	{
+		if elems[3].Elems() > 0 {
+			i.Committed = &Signature{}
+			if err := i.Committed.UnmarshalRLPWith(elems[3]); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// createValidatorSetDelta calculates ValidatorSetDelta based on the provided old and new validator sets
+func createValidatorSetDelta(log hclog.Logger, oldValidatorSet, newValidatorSet AccountSet) *ValidatorSetDelta {
+	var addedValidators AccountSet
+
+	oldValidatorSetMap := make(map[types.Address]*ValidatorAccount)
+	removedValidators := map[types.Address]int{}
+
+	for i, validator := range oldValidatorSet {
+		if (validator.Address != types.Address{}) {
+			removedValidators[validator.Address] = i
+			oldValidatorSetMap[validator.Address] = validator
+		}
+	}
+
+	for _, newValidator := range newValidatorSet {
+		// Check if the validator is among both old and new validator set
+		oldValidator, ok := oldValidatorSetMap[newValidator.Address]
+		if ok {
+			if !oldValidator.Equals(newValidator) {
+				log.Error(
+					fmt.Sprintf(
+						"validator '%s' found in both old and new validator set, but its BLS keys differ",
+						newValidator.Address.String(),
+					),
+				)
+			}
+
+			// If it is, then discard it from removed validators...
+			delete(removedValidators, newValidator.Address)
+		} else {
+			// ...otherwise it is added
```

Comment:
Comment adds no value

Reply:
To me, it is important to stress out that logic is "inverse" in a way that all the validators from the old validator set are initially marked as removed and as soon as it is found in the new validator set it gets discarded from the removed validators. Therefore I think we should have some comment outlining this. 

We are open for suggestion on how that comment should look like if you don't find the current one suitable (enough).